docs/proposals/multiple-sources-for-applications.md

```diff
@@ -0,0 +1,225 @@
+* https://github.com/argoproj/argo-cd/issues/2789
+
+## Open Questions
+* Adding external sources to the Application reource would add additional latencies for creation/reconcilation process. Should we add it to the doc in Risks?
+* 
+
+## Summary
+
+Currently, Argo CD supports Applications with only a single ApplicationSource. In certain scenarios, it would be useful to support more than one source for the application. For example, consider a need to create multiple applications with the same manifests. Today we would have to copy manifest files from one application to another.
+
+Creating applications from multiple sources would allow users to configure multiple services stored at various sources within the same application.
+
+## Motivation
+
+The main motivation behind this enhancement proposal is to allow users to create an application using services that are stored in various sources.
+
+### Goals
+
+The goals of the enhancement are:
+
+#### **Supporting multiple sources for creating an application**
+
+Users should be able to specify multiple sources to add services to the application. Argo CD should compile all the sources and reconcile each source individually for creating the application.
+
+#### **Allow specifying external value files for Helm repositories**
+
+Users should be able to specify different sources for Helm charts and values files. The Helm charts specified by the user could be available in Git or Helm repository and the value files are stored in Git. Argo CD should track changes in both the Helm charts and the value files repository and reconcile the application.
+
+#### Changes to UI
+
+The UI should allow users to add multiple sources while creating the application. For each source, UI should allow users to add multiple external values files Helm projects.
+
+#### Changes to cli
+
+The cli would need to support adding a list of resources instead of just one while creating the application. Also, just like `--values` field for adding values files, we would need an option to allow external value files to the application.
+
+### Non-goals
+* Allow reconciliation from Argo CD Application resources that are located at various sources. We believe this would be possible with some adaptions in the reconcilation workflow.
+
+## Proposal
+
+### Add new `sources` field in Application Spec
+
+The proposal is to add a new field `sources` which would allow users to input list of `ApplicationSource`s. We would mark field `source` as deprecated and would override the details under `source` with details under `sources` field.
+
+```yaml
+spec:
+  source:
+    repoURL: https://charts.bitnami.com/bitnami
+    targetRevision: 8.5.8
+    helm:
+    valueFiles:
+        - values.yaml
+    chart: mysql
+  sources:                                          # new field
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+      chart: mysql
```

Comment:
Do we need a `name` field to disambiguate in cases where the chart or repo name is the same for two sources? 

For example, it seems like this could be a problem:

```yaml
  sources:
    - repoURL: https://charts.bitnami.com/bitnami
      targetRevision: 8.5.8
      helm:
        valueFiles:
          - values.yaml
      chart: mysql
    - repoURL: https://my.company.com/org/mysql.git
      targetRevision: HEAD
      path: '.'
```

I guess when we clone the repo, we could suffix the target directory with the list index (e.g. `mysql-0`). But if we were to allow cross-source communication (e.g. `valueFiles: [../mysql/values.yaml]`), that convention might not be obvious/intuitive.

Reply:
I think having a unique identifier (maybe `name` field) would help in scenarios if we want to include cross-source resource sharing. I am not able to come up with other use-cases for the `name` field at the moment.